File: medium/can_be_prepared.go
package medium

import (
	core "github.com/wearypossum4770/dark-coding-challenges/dark_coding_challenges"
	"strings"
)

func CanBePrepared(recipe string, ingredients []byte) bool {
	if len(recipe) == 0 || len(ingredients) == 0 {
		return false
	}

	outer, inner, set := []bool{}, []byte{}, core.CreateSetLike(ingredients)

	recipe = strings.ReplaceAll(recipe, " ", "")
	for i := range recipe {
		char := recipe[i]
		if core.IsAlpha(char) {
			inner = append(inner, char)
			continue
		}
		switch char {
		case '|', '(':
			outer = append(outer, core.AllTrue(core.SetContains(set, inner)))
			inner = inner[:0]
		}
	}
	if len(inner) > 0 {
		outer = append(outer, core.AllTrue(core.SetContains(set, inner)))
	}
	return core.AnyTrue(outer)
}

File: medium/can_be_prepared.py
def can_be_prepared(recipe: str, ingredients: tuple[str, ...]) -> bool:
    if not recipe or not ingredients:
        return False
    outer = []
    inner = []
    for piece in recipe:
        if piece.isalpha():
            inner.append(piece)
        elif piece in "|(" and inner:
            outer.append(all([c in ingredients for c in inner]))
            inner = []
    if inner:
        outer.append(all([c in ingredients for c in inner]))
    return any(outer)
